plugins/askrene/algorithm.c

```diff
@@ -341,68 +341,329 @@ s64 node_balance(const struct graph *graph,
 	return balance;
 }
 
+/* Helper.
+ * Compute the reduced cost of an arc. */
+static inline s64 reduced_cost(const struct graph *graph, const struct arc arc,
+			       const s64 *cost, const s64 *potential)
+{
+	struct node src = arc_tail(graph, arc);
+	struct node dst = arc_head(graph, arc);
+	return cost[arc.idx] - potential[src.idx] + potential[dst.idx];
+}
+
+/* Finds an optimal path from the source to the nearest sink node, by definition
+ * a node i is a sink if node_balance[i]<0. It uses a reduced cost:
+ *	reduced_cost[i,j] = cost[i,j] - potential[i] + potential[j]
+ *
+ * */
+static struct node dijkstra_nearest_sink(const tal_t *ctx,
+					 const struct graph *graph,
+					 const struct node source,
+					 const s64 *node_balance,
+					 const s64 *capacity,
+					 const s64 cap_threshold,
+					 const s64 *cost,
+					 const s64 *potential,
+					 struct arc *prev,
+					 s64 *distance)
+{
+	struct node target = {.idx = INVALID_INDEX};
+	tal_t *this_ctx = tal(ctx, tal_t);
+
+	if (!this_ctx)
+		/* bad allocation */
+		goto finish;
+
+	/* check preconditions */
+	assert(graph);
+	assert(node_balance);
+	assert(capacity);
+	assert(cost);
+	assert(potential);
+	assert(prev);
+	assert(distance);
+
+	const size_t max_num_arcs = graph_max_num_arcs(graph);
+	const size_t max_num_nodes = graph_max_num_nodes(graph);
+
+	assert(source.idx < max_num_nodes);
+	assert(tal_count(node_balance) == max_num_nodes);
+	assert(tal_count(capacity) == max_num_arcs);
+	assert(tal_count(cost) == max_num_arcs);
+	assert(tal_count(potential) == max_num_nodes);
+	assert(tal_count(prev) == max_num_nodes);
+	assert(tal_count(distance) == max_num_nodes);
+
+	for (size_t i = 0; i < max_num_arcs; i++) {
+		/* is this arc saturated? */
+		if (capacity[i] < cap_threshold)
+			continue;
+
+		struct arc arc = {.idx = i};
+		struct node tail = arc_tail(graph, arc);
+		struct node head = arc_head(graph, arc);
+		s64 red_cost =
+		    cost[i] - potential[tail.idx] + potential[head.idx];
+
+		/* reducted cost cannot be negative for non saturated arcs,
+		 * otherwise Dijkstra does not work. */
+		if (red_cost < 0)
+			goto finish;
+	}
+
+	for (size_t i = 0; i < max_num_nodes; ++i)
+		prev[i].idx = INVALID_INDEX;
+
+/* Only in debug mode we keep track of visited nodes. */
+#ifndef NDEBUG
+	bitmap *visited =
+	    tal_arrz(this_ctx, bitmap, BITMAP_NWORDS(max_num_nodes));
+	assert(visited);
+#endif
```

Comment:
We don't generally use NDEBUG in our code (it turns off your assert()s above, too).  So this probably needs to be a runtime flag, like dev_askrene_debug, and maybe make it a global set if --developer is true?

Reply:
OK. Then I will remove this check which is unnecessary.